Marlin/src/module/stepper.cpp

```diff
@@ -1506,131 +1407,230 @@ void Stepper::isr() {
     #if EXTRA_CYCLES_XYZE > 20
       if (i) while (EXTRA_CYCLES_XYZE > (uint32_t)(HAL_timer_get_count(PULSE_TIMER_NUM) - pulse_start) * (PULSE_TIMER_PRESCALE)) { /* nada */ }
     #elif EXTRA_CYCLES_XYZE > 0
-      if (i) DELAY_NOPS(EXTRA_CYCLES_XYZE);
+      if (i) DELAY_NS(EXTRA_CYCLES_XYZE * NANOSECONDS_PER_CYCLE);
     #endif
 
   } // steps_loop
+}
 
-  // Calculate new timer value
-  if (step_events_completed <= (uint32_t)current_block->accelerate_until) {
+// This is the last half of the stepper interrupt: This one processes and
+// properly schedules blocks from the planner. This is executed after creating
+// the step pulses, so it is not time critical, as pulses are already done.
 
-    #if ENABLED(BEZIER_JERK_CONTROL)
-      // Get the next speed to use (Jerk limited!)
-      hal_timer_t acc_step_rate =
-        acceleration_time < current_block->acceleration_time
-          ? _eval_bezier_curve(acceleration_time)
-          : current_block->cruise_rate;
-    #else
-      #ifdef CPU_32_BIT
-        MultiU32X24toH32(acc_step_rate, acceleration_time, current_block->acceleration_rate);
+uint32_t Stepper::stepper_block_phase_isr() {
+
+  // If no queued movements, just wait 1ms for the next move
+  hal_timer_t interval = (HAL_STEPPER_TIMER_RATE / 1000);
+
+  // If there is a current block
+  if (current_block) {
+
+    // Calculate new timer value
+    if (step_events_completed <= current_block->accelerate_until) {
+
+      #if ENABLED(BEZIER_JERK_CONTROL)
+        // Get the next speed to use (Jerk limited!)
+        hal_timer_t acc_step_rate =
```

Comment:
If `Stepper:acc_step_rate` is being changed to always be `uint32_t` then perhaps this too?

Reply:
I think you are right! 👍

Reply:
The reason for changing everything to uint32_t is that on AVR we emulate a 32bit counter (previously, it was also done by splitting interrupts) and on ARM timers are 32bits at least... ;)

Reply:
In Stepper.cpp, there are 3 places where that should be changed: